### DISCUSSION TITLE
How do I load all documents in a (nested) directory structure

### DISCUSSION THREAD

**Discussion by BaukeL** [eclipse-langium/langium#1778] [Q&A] [answered]
In the snippet below only one document is added tot the workspace. How can I do something similar by providing a directory name? Based on this directory name all files with applicable extensions should then give a list of documents to be passed to the DocumentBuilder. Or is there another way this should work?

```
export async function extractDocument<T extends AstNode>(fileName: string, extensions: readonly string[], services: LangiumCoreServices): Promise<LangiumDocument<T>> {

//....
    const document = await services.shared.workspace.LangiumDocuments.getOrCreateDocument(URI.file(path.resolve(fileName)));
    await services.shared.workspace.DocumentBuilder.build([document], { validation: true });
//....
}
```

**Comment by msujew** [comment, marked as answer]
Hey @BaukeL,

there are two ways to go about this:
1. As you've outlined, iterate through all the files in a directory via the `fs` import and create documents for all of them and pass them into the `DocumentBuilder.build` method.
2. Use the `WorkspaceManager#initializeWorkspace` method and pass in an array of `WorkspaceFolder`. It calls `DocumentBuilder.build` internally. See:
https://github.com/eclipse-langium/langium/blob/3f9de4cf277a95d4dd525dab2e8eecfb7cce6044/examples/domainmodel/src/cli/cli-util.ts#L45-L57
